nion/swift/test/ImageCanvasItem_test.py

```diff
@@ -291,6 +291,27 @@ def test_hand_tool_on_one_image_of_multiple_displays(self):
             document_controller.tool_mode = "hand"
             display_panel.display_canvas_item.simulate_press((100,125))
 
+    def test_zoom_tool_on_one_image_of_multiple_displays(self):
```

Comment:
Can you rename this to something like `test_zoom_in_and_out_around_clicked_point`? The `one_image_of_multiple_displays` is spurious.

Reply:
Test has been renamed, and the mag glass reference removed.
The zooming maintaining the cursor position has been sorted out, that function that you mentioned was a bit of a red herring but I could jiggle around my mapping of widget to image coordinate spaces anyway to get what was needed for the extra vector needing to be applied.

On the single button option, that isn't quite what I thought was agreed as things seem to reference buttons plural. Changing it to a single zoom button would make it pretty opaque on how to use it (needing to press random buttons to find out how to zoom out). It also begs the question about the two icons you sent me to use :)  For the time being it would make sense to go with the two button option, accept that we know it isn't perfect but it is still better than not having buttons at all. With the way the discussions were going anyway, the buttons are likely to be irrelevant entirely and simply be replaced with using the scroll-wheel on the mouse and the buttons being a backup.

On the 'z' shortcut key, personally I am not a fan of shortcut keys that can be accidently activated as it can lead to confusion. I know you have some existing shortcut keys that are single key ones and if that is your preferred pattern then for the future block of work we could add them in but again it may become redundant with the scroll wheel.

Reply:
Thanks for renaming the test. Can you add code to the test to ensure that it is actually zooming around the clicked point? I think there are similar examples in the `ImageCanvasItem_test.py` file, but if you need specific references, let me know.

Regarding the single button for the tool: zooming in/out on the same point is often required to get the view just right; having two tools requires moving the mouse back to the tool bar to switch, losing the zoom point in the process. As far as I can see from a quick search, the industry norm is a single tool in software that uses tools. Some software allows you to place a point and zoom in/out on that point with a keybaord. We can consider that in the future, i.e. "zoom around graphic". I couldn't find another program that has two zoom tools. The option and control keys already modify many of the other tools, and it's an industry standard, so I don't think it's a big stretch for users to discover the functionality. We can add a cursor that changes in the future to make it easier. I'd also like double clicking the zoom tool to reset the display to 'fit'. And eventually dragging to fill the display panel with the dragged area.

I personally don't like scroll-wheel-zooming, but we can try it out in another PR. However, I think we should spend our time on this PR to get it right, without assuming it will be replaced by something else in the future.

Regarding the shortcut keys - I'm open to different ideas. I'm not a fan of the single keys affecting modes either. Maybe we do something like multiple keys? Esc+'z' to switch modes? We definitely will need some sort of key shortcut for this and other tools. Let's open a discussion offline.

The `mag_glass_in`/`mag_glass_out` files are spurious. And I've attached a replacement for the magnifying glass.

![zoom](https://github.com/nion-software/nionswift/assets/34664/7720dd56-d01f-48b9-aee0-4d0cee760586)
![zoom](https://github.com/nion-software/nionswift/assets/34664/ca1cd8bb-d4db-484b-9788-f58a2119965e)

Reply:
I can certainly look to add extra checks to the test, though I have a query as to what these tests are trying to actually verify.  I mirrored the 'hand tool' test, almost exactly so I am trying to understand what these test are attempting to validate. The 'hand tool' test as far as I can see validates that the code path is runnable (as do other tests in the ImageCanvasItem_test file), with no defined asserts so I was working on the idea that the tests validate all code paths are runnable, with extra asserts added to validate state easily.  If instead the objective is to try and go full TDD, then surely we need the test requirements pre-defined as part of the Issue definition which can then be used as verification of completion of functionality.

Shortcut-keys, they tend to be either Ctrl-X or Alt-X as holding that extra key is something explicit the user has to do and is unlikely to be caught by accident.  My concern with just the single keys is it is quite easy to knock one by accident and suddenly whichever image you had selected may/may not have changed in some unforeseen way. If it is consistent with the way Swift works then we can add it in, just trying to work through the 'question everything' phase to understand why.

Reply:
On the single Zoom icon, I can certainly change it to a single icon instead and we can go off left-click/right-click as the convention on quite a bit of software seems to be. This especially seems to be the case in more modern software to go with a single icon rather than two, in parallel with the scroll-to-zoom which now seems to be everywhere.  My reasoning for pushing back is the original discussion was always about buttons plural, and you yourself sent me two icons, so now with it suddenly switching to wanting a single icon I am trying to understand the reasoning for the change both from a 'are we building the right thing' but also a process point of view. If it is simply that the original requirements were wrong, and you hadn't thought about the implications of the icons you were sending, then that is ok but trying to make sure we stay on the same page here.

Since scroll-to-zoom wasn't included in the original Issue as a concrete point, just some discussion I am going to park those for the moment until there is consensus on it. Similarly with the 'zoom around graphic', double-click to fit, drag-to-zoom (though code-behind for this is already done, minus the drawing of the selection rectangle).  If we want to go whole-hog on this Pull Request and fix everything to do with zooming with it, then I am ok with that but we should recognise that we are somewhat deviating from the process of fixing a well defined Issue and going with a more fluid 'experimental' UX development.

Reply:
I'm not proposing to expand this PR in any way - just to implement it in a reasonable way. Zoom tool that zooms in/out with adequate test. No key overload (for now). No dragging (for now). No zoom around graphic (for now). No zoom-specific cursor (for now).

The specific test to copy is this one: `test_hand_tool_undo` in `DisplayPanel_test.py`. Please put it in `DisplayPanel_test.py`, even though that might not be the ideal spot. Its good to keep the two tests together.

Regarding the tool: here's the gold standard for the zoom tool in applications like this:
https://helpx.adobe.com/illustrator/using/tool-techniques/zoom-tool.html

I've never seen a program use click/right-click in a zoom button. But either way, right-click is overloaded to do the context menu in Swift. Also, track pads (laptops) click/right-click would be burdensome. Let's just follow the gold standard linked above.

Originally I provided two icons because I didn't think through the issue. After trying it out, it is really obvious that we should have a single zoom tool that switches when holding the Alt/Opt key.